server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health.node;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.health.Diagnosis;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class DiskHealthIndicatorService implements HealthIndicatorService {
+    public static final String NAME = "disk";
+
+    private static final Logger logger = LogManager.getLogger(DiskHealthIndicatorService.class);
+
+    private final ClusterService clusterService;
+
+    public DiskHealthIndicatorService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean explain, HealthInfo healthInfo) {
```

Comment:
I find this method difficult to read and reason about. 

I am okay if we keep it like this for now, but I would like to go over it one more time and either restructure it a bit or add comments. I think it will be really difficult for someone that hasn't worked on this as extensively as we to maintain it in the future.

If you agree I will create a github issue. Let me know, if it's just me.

Reply:
I find it difficult to reason about it too.

Hopefully, if we update the behaviour the design will crystalise too ( https://github.com/elastic/elasticsearch/pull/90041#discussion_r975144475 ).

Reply:
I think you both found it difficult to follow because we had a miscommunication about the expected behavior. I had thought @andreidan only wanted information about the highest-priority problem (I thought we'd gone through that logic offline). I've changed it now to return information about all problems on all node types (data, master, or other) at all levels (red or yellow).